drivers: mspi: mspi_dw: fixes to work on additional platforms

danieldegrasse opened the pull request:
This PR includes a series of fixes to support the MSPI_DW driver on additional platforms, beyond the NRF SOC where it was initially enabled. I've split the changes into logical commits within this PR, and can split the PR into multiple parts if desired.

Below is a summary of the changes (ordered by how impactful I think they are):
- **Change bit offsets for CTRLR0 register in driver**- I am not sure if these offsets differ between IP revisions, @anangl can you check this?
- Add `mspi_timing_cfg` implementation to configure the `RX_SAMPLE_DLY` register (used for adjusting RX sample delay at high clock speeds)
- Write the DFS32 mask as well as DFS, since some IP revisions use this field to set dataframe size
- read the ICR bit at the exit of the ISR handler, to clear pending ISR signals
- Enable support for multi-level interrupts in the driver
- add generic vendor specific functions (for cases when the functions can simply be a NOP)
- properly support compiling the driver when `CONFIG_PINCTRL=n`

danieldegrasse committed:
drivers: mspi: supply empty vendor specific macros for generic case

Supply empty vendor specific macros for cases where vendor specific
functions are not needed by driver.

Signed-off-by: Daniel DeGrasse <ddegrasse@tenstorrent.com>

danieldegrasse committed:
drivers: mspi: mspi_dw: make pinctrl include conditional

The remainder of this driver functions when pin control is disabled,
so make the include conditional as well to fully support this case.

Signed-off-by: Daniel DeGrasse <ddegrasse@tenstorrent.com>

danieldegrasse committed:
drivers: mspi: mspi_dw: support newer IP revisions

Bitmask offsets for the MSPI_DW peripheral are incorrect for the
4.03a databook for the DW APB SSI. Add a "v2" compatible to handle these
changed offsets. The compatible does not define new binding properties,
just modifies the register offsets.

Additionally, handle writing the DFS32 mask for v2 of this compatible,
as newer IP supports a dataframe size up to 32 bits.

Signed-off-by: Daniel DeGrasse <ddegrasse@tenstorrent.com>

danieldegrasse committed:
drivers: mspi: mspi_dw: properly support multi-level interrupts

Multilevel interrupts require that the irq number be read with DT_IRQN,
so update the driver to use this macro.

Signed-off-by: Daniel DeGrasse <ddegrasse@tenstorrent.com>

danieldegrasse committed:
drivers: mspi: mspi_dw: add API to configure RX_DLY timing

The SSI DW peripheral supports an RX_SAMPLE_DLY register in some
instances- this register controls the number of clock cycles from the
default sample time before the RX input is actually sampled. This can be
used to improve reliability when operating the SSI at a higher clock
speed.

Add an implementation of the mspi_timing_cfg api, and header to define
the identifier so that users can configure this parameter

Signed-off-by: Daniel DeGrasse <ddegrasse@tenstorrent.com>

anangl commented on drivers/mspi/mspi_dw.h:
```diff
@@ -10,23 +10,23 @@
 
 /* CTRLR0 - Control Register 0 */
-#define CTRLR0_SPI_FRF_MASK	GENMASK(23, 22)
+#define CTRLR0_SPI_FRF_MASK	GENMASK(22, 21)
```
> - **Change bit offsets for CTRLR0 register in driver**- I am not sure if these offsets differ between IP revisions, @anangl can you check this?

Apparently they differ. I based on the 1.03a databook, as that's the SSI revision used in the nRF EXMIF peripheral. I've checked also the 2.00a databook, and the offsets are the same there; they must have been changed later. The driver would need to support both sets then. I think it should be fine to select the proper one at build time, instead of using the SSIC_COMP_VERSION register for this.

danieldegrasse replied:
Thanks for checking this, I figured the offsets would differ. I have version 4.03a. I'd argue we should use two bindings to distinguish the difference here, similar to how linux does: https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/spi/snps%2Cdw-apb-ssi.yaml#L55

Maybe `snps_designware_ssi` and `snps_designware_ssi-v2`? We could also use `snps_designware_ssi-4.03a` based off the databook version

anangl replied:
Yes, a different compatible string seems to be a good solution. I think I'd rather go with `-v2`, as we don't know if this change was introduced exactly in version 4.03a.

anangl commented on drivers/mspi/mspi_dw.c:
```diff
@@ -307,7 +308,9 @@ static void mspi_dw_isr(const struct device *dev)
 		k_sem_give(&dev_data->finished);
 	}
 
+	read_icr(dev);
```
Is this actually needed? The flags used for interrupts generation (RXFIS and TXEIS) are cleared by hardware when the FIFOs are read/written, respectively.

> Per the SSI DW documentation, the ICR bit should be read within an
interrupt to clear pending interrupts. Add this to the ISR handler so
interrupts are not fired continuously.

I couldn't find such statement in the documentation that the register should be read. It can be read to clear the interrupts, but I don't think there's an obligation to do so.

danieldegrasse replied:
It appears not to be- I'm guessing I misunderstood the documentation. SPI driver still seems to work without this change locally, so I'll drop it.

anangl commented on drivers/mspi/mspi_dw.h:
```diff
@@ -27,6 +27,7 @@
 #define CTRLR0_FRF_SSP		1UL
 #define CTRLR0_FRF_MICROWIRE	2UL
 #define CTRLR0_DFS_MASK		GENMASK(4, 0)
+#define CTRLR0_DFS32_MASK	GENMASK(20, 16)
```
This field should be added and used conditionally, based on `SSI_VERSION_2`, as it is not present e.g. in revision 1.03a -bits 19:16 are used for CFS (Control Frame Size, for Microware) field and bits 21:20 are reserved there.

danieldegrasse replied:
Fixed to define the mask as 0x0 when SSI_VERSION_2 is not defined- does this work?

anangl commented on drivers/mspi/mspi_dw.c:
```diff
@@ -1379,16 +1379,17 @@ static DEVICE_API(mspi, drv_api) = {
 	.dev_config         = api_dev_config,
 	.get_channel_status = api_get_channel_status,
 	.transceive         = api_transceive,
+	.timing_config      = api_timing_config,
```
This should not be added in the "drivers: mspi: mspi_dw: properly support multi-level interrupts" commit.

danieldegrasse replied:
Good catch, should be fixed

anangl commented on drivers/mspi/mspi_dw.h:
```diff
@@ -9,24 +9,34 @@
  * It is not supposed to be included by any file other than mspi_dw.c.
  */
 
+#include <zephyr/sys/util.h>
```
This is included by `mspi_dw.c` before `mspi_dw.h`.

danieldegrasse replied:
Good point-removed

anangl commented on drivers/mspi/mspi_dw.c:
```diff
@@ -1223,6 +1236,17 @@ static int _api_xip_config(const struct device *dev,
+			     const uint32_t param_mask, void *cfg)
+{
+	if (param_mask & MSPI_DW_RX_TIMING_CFG) {
+		write_rx_sample_dly(dev, (uint32_t)(uintptr_t)cfg);
```
Isn't the `cfg` parameter intended to point to a structure? What if another timing parameter was to be added to be configurable for the driver?

swift-tk replied:
Yes, it is best for void *cfg to point to a structure that contains all dw specific timing parameters, so that you could get it from devicetree all at once and apply them all. But again, the timing_config API is meant to be flexible and vendor specific, so you could decide to code whatever way you like..

danieldegrasse committed:
drivers: mspi: mspi_dw: improve reliability of CS signal

The designware controller has an *interesting* implementation of the CS
signal- CS will be de-asserted whenever the TX FIFO is empty, so slower
cores may see CS de-assert prematurely if they cannot keep pace with
their SPI peripheral. To help reduce odds of de-assertion, implement the
following changes:

- don't write SER bit until directly before we enable interrupts, so
  that transfers don't start early
- prefix the TX FIFO before writing SER, so the FIFO can drain a bit
  before have to service an interrupt

Signed-off-by: Daniel DeGrasse <ddegrasse@tenstorrent.com>

danieldegrasse committed:
drivers: mspi: mspi_dw: support DDR mode

Support DDR mode within the mspi_dw driver

Signed-off-by: Daniel DeGrasse <ddegrasse@tenstorrent.com>

danieldegrasse committed:
drivers: mspi: mspi_dw: make sure controller is disabled at entry

Some systems (IE cores with bootloaders) may not disable the DW SPI
peripheral at boot time. Make sure the peripheral is disabled before we
try to configure it.

Signed-off-by: Daniel DeGrasse <ddegrasse@tenstorrent.com>

danieldegrasse commented:
Latest push also includes support for DDR mode, since that is used by our platform. I can split that into another PR if desired.

danieldegrasse commented:
@anangl can you revisit when you have a chance?

anangl reviewed (commented):
@danieldegrasse Apologies that it took me that long to get back to this.

danieldegrasse commented:
Updated to address comments- I also added locking within the `timing_config` API function, since other APIs use locks before configuring the MSPI

anangl commented on drivers/mspi/mspi_dw.c:
```diff
@@ -994,8 +993,17 @@ static int start_next_packet(const struct device *dev, k_timeout_t timeout)
 		}
 	}
 
+	/* Prefill TX FIFO with any data we can */
+	if (dev_data->dummy_bytes && tx_dummy_bytes(dev)) {
+		imr = IMR_RXFIM_BIT;
+	} else if (packet->dir == MSPI_TX && packet->num_bytes) {
+		tx_data(dev, packet);
+	}
```
> If we start transfers without the TX FIFO filled, the interrupt latency will cause it to drain before the first ISR is serviced and CS will deassert.

This won't happen because the start level is set to maximum. Thus, the transmission will not start until the first ISR fills up the TX FIFO. See this part of code: https://github.com/zephyrproject-rtos/zephyr/blob/5ac8c48a0a73e96f261ed659785463924492331a/drivers/mspi/mspi_dw.c#L941-L987

anangl commented on drivers/mspi/mspi_dw.c:
```diff
@@ -1223,6 +1263,37 @@ static int _api_xip_config(const struct device *dev,
+		}
+
+		(void)k_sem_take(&dev_data->ctx_lock, K_FOREVER);
+		write_rx_sample_dly(dev, config->rx_sample_dly);
```
I think it would be better to just store this value in `dev_data` here and write the register in `start_next_packet()`, together with the others that cannot be written when the controller is enabled (databook 1.03a says that's the case for this register, too).

anangl commented on drivers/mspi/mspi_dw.c:
```diff
@@ -994,8 +993,17 @@ static int start_next_packet(const struct device *dev, k_timeout_t timeout)
 	/* Enable interrupts now and wait until the packet is done. */
 	write_imr(dev, imr);
+	/* Write SER to start transfer */
+	write_ser(dev, BIT(dev_data->dev_id->dev_idx));
```
I tried this on v1.03a and it seems to work fine, so perhaps nothing has changed in the behavior but just the sentence that I cited was not phrased properly (actually, how could the SSI be busy without being enabled?).

cfriedt commented on drivers/mspi/mspi_dw.c:
```diff
@@ -912,6 +926,12 @@ static int start_next_packet(const struct device *dev, k_timeout_t timeout)
 	write_rx_sample_dly(dev, dev_data->rx_sample_dly);
+	if (dev_data->spi_ctrlr0 & (SPI_CTRLR0_SPI_DDR_EN_BIT |
+				    SPI_CTRLR0_INST_DDR_EN_BIT)) {
+		write_txd_drive_edge(dev, dev_data->baudr / 4);
```
@danieldegrasse - are you able to address this comment? I think it came in after approval. @anangl - do you need this resolved before this PR is merged?

danieldegrasse replied:
Ah sorry just saw this- I can make a follow up PR to support configuring this.

danieldegrasse replied:
https://github.com/zephyrproject-rtos/zephyr/pull/94088

utsavm9 commented on drivers/mspi/mspi_dw.c:
```diff
@@ -1231,6 +1235,20 @@ static int _api_xip_config(const struct device *dev,
 	return 0;
 }
 
+static int api_timing_config(const struct device *dev,
```
I believe @danieldegrasse you added this behind `CONFIG_MSPI_XIP` when it was not intended to be such.
